Source/Core/VideoCommon/OnScreenDisplay.cpp

```diff
@@ -110,10 +117,12 @@ void DrawMessages()
 
   for (auto it = s_messages.begin(); it != s_messages.end();)
   {
-    const Message& msg = it->second;
+    Message& msg = it->second;
     const int time_left = static_cast<int>(msg.timestamp - now);
 
-    if (time_left <= 0)
+    // Make sure we draw them at least once if they were printed with 0ms,
+    // unless enough time has expired, in that case, we drop them
+    if (time_left <= 0 && (msg.ever_drawn || -time_left >= MESSAGE_DROP_TIME))
```

Comment:
Instead of this `ever_drawn` variable which means a couple locations have to be non-const, why not always draw the message first before seeing if it needs to be removed?

Could use a helper function to make it a little cleaner.

Ex:

```
void DrawMessages()
{
    const bool draw_messages = Config::Get(Config::MAIN_OSD_MESSAGES);
    if (!draw_messages)
      return;
    
  const u32 now = Common::Timer::GetTimeMs();
  const float current_x =
      LEFT_MARGIN * ImGui::GetIO().DisplayFramebufferScale.x + s_obscured_pixels_left;
  float current_y = TOP_MARGIN * ImGui::GetIO().DisplayFramebufferScale.y + s_obscured_pixels_top;
  int index = 0;
  std::lock_guard lock{s_messages_mutex};
  
    for (auto it = s_messages.begin(); it != s_messages.end();)
    {
      const Message& msg = it->second;
      const int time_left = static_cast<int>(msg.timestamp - now);
      DoDrawMessage(time_left, msg, index, current_x, current_y);
      
      if (time_left <= 0)
      {
        it = s_messages.erase(it);
      }
      else
      {
        ++it;
      }
    }
}

void DoDrawMessage(int time_left, const Message& msg, int& index, float current_x, float& current_y)
{
  if (-time_left >= MESSAGE_DROP_TIME)
    return;

  current_y += DrawMessage(index++, msg, ImVec2(current_x, current_y), time_left);
}
```

Reply:
That would mean that all the expired messages draw one time more than necessary (which also mean they wouldn't follow the length of time you had specified). I think it's better to have a variable non const.

Reply:
I agree with you about your concerns with my suggestion.  Still not a fan of that variable.  What do you think about this - add another timestamp `s_last_drawn_timestamp`.  Then your if check can be changed slightly:

```
if (time_left <= 0 && (msg.timestamp <= s_last_drawn_timestamp || -time_left >= MESSAGE_DROP_TIME))
```

At the end of the function:

```
s_last_drawn_timestamp = now;
```

Reply:
Hypothetical case but that would break when going over ~1000fps though (time accuracy is 1ms):
Add a message: timestamp 53 (53ms current time + 0 duration).
Draw messages at time 53ms.
Add another message which is exactly the same: timestamp 53 (if we are still within the same ms).
Draw messages at time 53ms again: the previous message nor the new message will draw now.

Reply:
That's true.  Sorry, my initial comment was not well thought out.  However, still not a fan of that variable so I have a different suggestion.

If I'm understanding things correctly, I think what you're trying to account for are messages that get missed because they expire sometime before `DrawMesages()` gets called (or in the case of a start/stop of the software expire sometime in between `DrawMessages()` calls).

I think a solution to that problem is to add another timestamp.

```
u32 s_last_drawn_timestamp = 0;

void DrawMessages()
{
  if (time_left <= 0 && (s_last_drawn_timestamp > msg.timestamp || -time_left >= MESSAGE_DROP_TIME))
  ...
  s_last_drawn_timestamp = now;
}
```

Reply:
Yeah with > instead of >= it might work. Thinking about it.

Reply:
changing >= to > has the opposite effect of the problem I mentioned above. If going to extremely high fps, they could now draw more than once. Which is not really a problem, but is theoretically wrong, and given that with the ever_drawn flag it works perfectly, I don't think it would be a worth changing.

Reply:
> If edits are requested for such small things none of my other changes will ever go it, not within 3+ years 

It's just an unfortunate consequence of this being an open-source repo with very few developers doing constant reviews.  Because of that, code changes can take a while.  Only thing that gets merged fast are code-cleanup changes and even that can sometimes take a while depending on the size.

>  If going to extremely high fps, they could now draw more than once.

I personally think that this is a fine trade-off for removing that variable (that is included on every message) and returning `const` correctness.  When you get to that high framerate, timing accuracy isn't going to matter that much.  If we really want to fix this, we need a more granular clock.  I'm going to guess we aren't using `std::chrono` in `Timer` because it didn't exist back then.

Reply:
Well on screen messages are not logs, there are barely any, what's the problem with adding a bool? And it's not like it's not const correct, ever_drawn is changed by the Draw() function. That kind of makes sense.

Reply:
I'm definitely nitpicking.  Sorry to take up so much of your time.  It just felt counterintuitive for `DrawMessage` to be changing `Message` even though I see your reason for doing so.  My hope was to come up with a solution that would allow for the `Message` to be `const` again.

Anyway, I don't have merge powers, so someone else will need to review this regardless.  My apologies again.  These are good changes overall.